Improve exception message when `readCsv` uses wrong `colType `

Repro, where a is `String`, but `ColType.Int` is used by accident:
```kt
val df = DataFrame.readCsvStr(
    """
    a
    abc
    """.trimIndent(),
    colTypes = mapOf("a" to ColType.Int),
)
```

Exception:
```
Exception in thread "main" java.lang.IllegalStateException: Could not read delimiter-separated data: CsvReaderException: Caught exception: Parsing failed on input, with nothing left to fall back to. Parser io.deephaven.csv.parsers.IntParser successfully parsed 0 items before failure.
	at org.jetbrains.kotlinx.dataframe.impl.io.ReadDelimDeephavenKt.readDelimImpl(readDelim.kt:168)
	at org.jetbrains.kotlinx.dataframe.io.ReadCsvStrKt.readCsvStr(readCsvStr.kt:51)
	at org.jetbrains.kotlinx.dataframe.io.ReadCsvStrKt.readCsvStr$default(readCsvStr.kt:33)
	at org.jetbrains.kotlin.MainKt.main(Main.kt:25)
	at org.jetbrains.kotlin.MainKt.main(Main.kt)
Caused by: io.deephaven.csv.util.CsvReaderException: Caught exception
	at io.deephaven.csv.reading.CsvReader.commonReadLogic(CsvReader.java:291)
	at io.deephaven.csv.reading.CsvReader.delimitedReadLogic(CsvReader.java:146)
	at io.deephaven.csv.reading.CsvReader.read(CsvReader.java:109)
	at org.jetbrains.kotlinx.dataframe.impl.io.ReadDelimDeephavenKt.readDelimImpl(readDelim.kt:155)
	... 4 more
Caused by: io.deephaven.csv.util.CsvReaderException: Parsing failed on input, with nothing left to fall back to. Parser io.deephaven.csv.parsers.IntParser successfully parsed 0 items before failure.
	at io.deephaven.csv.reading.ParseDenseStorageToColumn$OnePhaseParser.onePhaseParse(ParseDenseStorageToColumn.java:308)
	at io.deephaven.csv.reading.ParseDenseStorageToColumn$OnePhaseParser.access$000(ParseDenseStorageToColumn.java:297)
	at io.deephaven.csv.reading.ParseDenseStorageToColumn.doit(ParseDenseStorageToColumn.java:79)
	at io.deephaven.csv.reading.CsvReader.lambda$commonReadLogic$1(CsvReader.java:205)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:328)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:545)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:328)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1095)
```

--- a/dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/io/readDelim.kt
+++ b/dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/io/readDelim.kt
@@ -159,16 +159,24 @@ internal fun readDelimImpl(
                 /* sinkFactory = */ ListSink.SINK_FACTORY,
             )
         } catch (e: CsvReaderException) {
-            // catch case when the file is empty and header needs to be inferred from it.
-            if (e.message ==
-                "Can't proceed because hasHeaderRow is set but input file is empty or shorter than skipHeaderRows"
-            ) {
-                return@readDelimImpl DataFrame.empty()
+            when {
+                // catch case when the file is empty and header needs to be inferred from it.
+                e.message == FILE_EMPTY_EXCEPTION_MESSAGE ->
+                    return@readDelimImpl DataFrame.empty()
+
+                // provide extra information when an incorrect parser type is supplied
+                INCORRECT_PARSER_TYPE_EXCEPTION_MESSAGE in e.cause?.message.orEmpty() && colTypes.isNotEmpty() ->
+                    throw IllegalStateException(
+                        "Could not read delimiter-separated data. Check `colTypes`, you may have supplied the wrong type for a column. Cause: ${e.cause?.message}",
+                        e,
+                    )
+
+                else ->
+                    throw IllegalStateException(
+                        "Could not read delimiter-separated data: CsvReaderException: ${e.message}: ${e.cause?.message ?: ""}",
+                        e,
+                    )
             }
-            throw IllegalStateException(
-                "Could not read delimiter-separated data: CsvReaderException: ${e.message}: ${e.cause?.message ?: ""}",
-                e,
-            )
         }
     }
 
@@ -184,6 +192,12 @@ internal fun readDelimImpl(
 
     return dataFrameOf(cols)
 }
+
+private const val FILE_EMPTY_EXCEPTION_MESSAGE =
+    "Can't proceed because hasHeaderRow is set but input file is empty or shorter than skipHeaderRows"
+
+private const val INCORRECT_PARSER_TYPE_EXCEPTION_MESSAGE =
+    "Parsing failed on input, with nothing left to fall back to."
 
 @Suppress("UNCHECKED_CAST")
 private fun CsvReader.ResultColumn.toDataColumn(

--- a/dataframe-csv/src/test/kotlin/org/jetbrains/kotlinx/dataframe/io/DelimCsvTsvTests.kt
+++ b/dataframe-csv/src/test/kotlin/org/jetbrains/kotlinx/dataframe/io/DelimCsvTsvTests.kt
@@ -5,8 +5,10 @@ import io.kotest.assertions.throwables.shouldNotThrowAny
 import io.kotest.assertions.throwables.shouldThrow
 import io.kotest.matchers.collections.shouldContainInOrder
 import io.kotest.matchers.nulls.shouldNotBeNull
+import io.kotest.matchers.should
 import io.kotest.matchers.shouldBe
 import io.kotest.matchers.shouldNotBe
+import io.kotest.matchers.string.shouldContain
 import kotlinx.datetime.LocalDateTime
 import org.intellij.lang.annotations.Language
 import org.jetbrains.kotlinx.dataframe.DataFrame
@@ -863,6 +865,19 @@ class DelimCsvTsvTests {
             '{"firstName":"Jane","lastName":"Doe"}'
             
             """.trimIndent()
+    }
+
+    @Test
+    fun `incorrect colTypes exception`() {
+        shouldThrow<IllegalStateException> {
+            DataFrame.readCsvStr(
+                """
+                a
+                abc
+                """.trimIndent(),
+                colTypes = mapOf("a" to ColType.Int),
+            )
+        }.message!!.shouldContain("Check `colTypes`")
     }
 
     companion object {

